lib/segment/src/vector_storage/chunked_mmap_vectors.rs

```diff
@@ -0,0 +1,195 @@
+
+use crate::common::mmap_ops::{
+    open_write_mmap, transmute_from_u8_to_array, transmute_to_u8, transmute_to_u8_array,
+};
+use crate::data_types::vectors::VectorElementType;
+use crate::entry::entry_point::{OperationError, OperationResult};
+
+#[cfg(test)]
+const DEFAULT_CHUNK_SIZE: usize = 4 * 1024 * 1024; // 4Mb
+
+#[cfg(not(test))]
+const DEFAULT_CHUNK_SIZE: usize = 128 * 1024 * 1024; // 128Mb
+
+const CONFIG_FILE_NAME: &str = "config.json";
+const STATUS_FILE_NAME: &str = "status.json";
+
+const MMAP_CHUNKS_PATTERN_START: &str = "chunk_";
+const MMAP_CHUNKS_PATTERN_END: &str = ".mmap";
+
+struct ChunkedMMapStatus {
+    len: usize,
+}
+
+#[derive(Serialize, Deserialize)]
+struct ChunkedMMapConfig {
+    chunk_size: usize,
+    dim: usize,
+}
+
+pub struct ChunkedMMapVectors<T> {
+    config: ChunkedMMapConfig,
+    status: ChunkedMMapStatus,
+    _phantom: PhantomData<T>,
+    chunks: Vec<MmapMut>,
+    status_mmap: MmapMut,
+    directory: PathBuf,
+}
+
+impl ChunkedMMapVectors<VectorElementType> {
+    fn status_file(directory: &Path) -> PathBuf {
+        directory.join(STATUS_FILE_NAME)
+    }
+
+    fn write_to_mmap<T>(mmap: &mut MmapMut, offset: usize, data: T) {
+        debug_assert!(offset + size_of::<T>() <= mmap.len());
+        let ptr = mmap.as_mut_ptr();
+        unsafe {
+            let byte_data = transmute_to_u8(&data);
+            std::ptr::copy_nonoverlapping(byte_data.as_ptr(), ptr.add(offset), byte_data.len());
+        }
+    }
+
+    fn write_array_to_mmap<T>(mmap: &mut MmapMut, offset: usize, data: &[T]) {
+        debug_assert!(offset + data.len() * size_of::<T>() <= mmap.len());
+        let ptr = mmap.as_mut_ptr();
+        unsafe {
+            let byte_data = transmute_to_u8_array(data);
+            std::ptr::copy_nonoverlapping(byte_data.as_ptr(), ptr.add(offset), byte_data.len());
+        }
+    }
+
+    fn read_from_mmap<T>(mmap: &MmapMut, offset: usize) -> &T {
+        debug_assert!(offset + size_of::<T>() <= mmap.len());
+        let ptr = mmap.as_ptr();
+        unsafe { &*(ptr.add(offset) as *const T) }
+    }
+
+    fn read_array_from_mmap<T>(mmap: &MmapMut, offset: usize, len: usize) -> &[T] {
+        debug_assert!(offset + len * size_of::<T>() <= mmap.len());
+        let byte_len = len * size_of::<T>();
+        let slice = &mmap[offset..offset + byte_len];
+        transmute_from_u8_to_array(slice)
+    }
+
+    fn ensure_status_file(directory: &Path) -> OperationResult<MmapMut> {
+        let status_file = Self::status_file(directory);
+        if !status_file.exists() {
+            {
+                let length = std::mem::size_of::<usize>() as u64;
+                let file = OpenOptions::new()
+                    .create(true)
+                    .write(true)
+                    .open(&status_file)?;
+                file.set_len(length)?;
+            }
+            let mut mmap = open_write_mmap(&status_file)?;
+            Self::write_to_mmap(&mut mmap, 0, 0usize);
```

Comment:
This is obsolete as `set_len` fills the file with `0`s. But I guess we can keep it in as an extra safety measure, as we don't want this to mess up.

If we do remove this, we should `flush` after `set_len`.

Or just use serialized JSON here as well. Rather than messing with bytes ourselves.

Reply:
this one should be updated on insert, don't want json here